### DISCUSSION TITLE
[V5] Conditional Relationship Definition

### DISCUSSION THREAD

**Discussion by remidosol** [adonisjs/core#1786] [Help] [answered]
Hello folks 👋 

I'm confused about defining relationships by my User model's role.

I've a **enum type** that contains **'admin', 'designer', 'staff'** elements and used for **User's role column type**.
I want to define relationships by these roles.
e.g. **if the role is 'designer', HasMany relationship will be defined between User and Design model.** 
And **if the role is 'staff', HasMany relationship will be defined between User and StaffImage model.**

Is there any solution except creating models by these roles? Thanks already.

**Comment by hashtd** [comment]
_First thought_ 
I think you have to create HasMany relationship at Model level for User.

at Controller/Business logic level you can check role to inset/update receptive relationship

also at api json serialization based on role you can only pull that of specific user role conditional relationship 

_On Second thought_

only possible (guessing/speculation) alternative is to use override `boot()` method & based on role use `setRelatedForMany` method to set relationship

https://github.com/adonisjs/lucid/blob/77600cc00eb2629eec5e02018f4c247b603f6cd5/src/Orm/Relations/HasMany/index.ts#L168

but here loading data from table row then using `setRelatedForMany()` will be ideal flow, since you need role to determine which relationship need to be initialize

**Reply by remidosol** [reply]
May you give me an example code for both of your thoughts (especially your second thought)?

**Comment by fairhat** [comment, marked as answer]
If you have a fixed amount of roles (lets say 3 in your case), create models for each role (admin, designer, staff) which includes specific data for each role
For every role, assign its related relationship (design, staffImage, etc.)

In your controllers:
Check for the role of user (admin, designer, staff) and then fetch its role-model instance and then fetch the related data to the role

Example:
User (Staff) -> StaffProfile -> StaffImage

User (Designer) -> DesignerProfile -> Design

If this is too much abstraction, you could just assign all relationships to the user model and only query each when needed.

**Reply by fairhat** [reply]
Example with Users and Profiles (in your case you would use Profiles for your Roles)  

https://preview.adonisjs.com/guides/model-relations/has-one

**Reply by remidosol** [reply]
Thanks, I think this will be useful.
